### PR TITLE
[RPS-159] CI migration: map keywords for taxonomy

### PR DESCRIPTION
**Pull request by ngta1**

### CODE REVIEW

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.*;
+import org.yaml.snakeyaml.Yaml;
+
+public class TaxonomyLookup {
+
+    private final ArrayList<String> taxonomyKeys = new ArrayList<>();
+    private final static String HIPPO_TAXONOMY_KEY = "hippotaxonomy:key";
+
+    public TaxonomyLookup()
+    {
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java]: I think our conventions are opening bracket on same line
`public TaxonomyLookup() {`

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        }
+
+        // Extract all hippotaxonomy:key from yaml file
+        ExtractKeys(map);
+    }
+
+    private HashMap ExtractKeys(HashMap map) {
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java]: Method names with lower case?
`extractKeys()`

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                // build up a list of all keys first in taxonomyLookup.
+                if (HIPPO_TAXONOMY_KEY.equals(key)) {
+                    if (!taxonomyKeys.contains(value)) { taxonomyKeys.add(value);}
+                }
+            }
+        }
+        return map;
+    }
+
+    public boolean IsKeyExist(String key) {
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java]: Method names with lower case?
`isKeyExist()`

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        setup();
+    }
+
+    private void setup() {
+
+        // Load taxonomy yaml from file
+        // Currently uses development taxonomy file, but upon migration will use
+        // the full taxonomy structure from Initial Import of Taxonomy
+        Yaml yaml = new Yaml();
+        File taxonomyYamlFile = new File("./repository-data/development/src/main/resources/hcm-content/content/taxonomies/publication_taxonomy.yaml");
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java]: Is it worth putting the path as a program argument with this as the default so we don't have to change the code when we run it for real?

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+        HashMap<String, String> map = null;
+
+        try {
+            map = yaml.load(new FileInputStream(taxonomyYamlFile));
+
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java]: I'd re-throw this as a RuntimeException so the application doesn't continue if the error occurs

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    private HashMap ExtractKeys(HashMap map) {
+
+        // Apply to all the sub-maps as well
+        Iterator it = map.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry pair = (Map.Entry)it.next();
+
+            if (pair.getValue() instanceof HashMap) {
+               map.put(pair.getKey(), ExtractKeys((HashMap)pair.getValue()));
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java]: Don't think you need to put the result of ExtractKeys into the map as it's not getting modified. Should just be able to do
`ExtractKeys((HashMap)pair.getValue());`

You can then make this method void and not return the map as we don't need to look at the return value

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            }
+            else {
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/hippo/TaxonomyLookup.java]: Can put the else on one line like:
`} else {`

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/nesstar/PublishingPackage.java [hunk 8]
```diff
@@ -56,6 +59,23 @@ public String getAbstract() {
...
     }
 
+    public List<String> getKeywords() {
+
+        // Get TaxonomyLookup
+        TaxonomyLookup taxonomyLookup = new TaxonomyLookup();
+
+        // Get the nesstar keywords, and map to hippo taxonomy keys
+        return keywordsXpath.evaluate(rootElement).stream()
+            .flatMap(element -> {
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/nesstar/PublishingPackage.java]: As discussed we can use `.map()` to get nesstarNameToKeyword and then `.filter()` to strip out the ones that don't exist.

**Comment by ngta1** [comment]
Good call.  Will update.

File: migrator/src/main/java/uk/nhs/digital/ps/migrator/model/nesstar/PublishingPackage.java [hunk 9]
```diff
@@ -56,6 +58,22 @@ public String getAbstract() {
...
+        return keywordsXpath.evaluate(rootElement).stream()
+            .map(element -> element.getValue()
+                    .trim()
+                    .toLowerCase()
+                    .replace(" ", "-")
+            )
+            .filter(element -> taxonomyLookup.isKeyExist(element))
+            .collect(Collectors.toList());
+    }
+
```
Comment (jacsamell) [inline migrator/src/main/java/uk/nhs/digital/ps/migrator/model/nesstar/PublishingPackage.java]: 👍

**Pull request merged**